inference_worker.py

```diff
@@ -116,24 +135,212 @@ def custom_getter(getter, name, *args, **kwargs):
+        """Device-specific wrapper around a call to _load_graph.
+
+        Must be called with self._lock held for write.
+        """
+        pass
+
+    @abc.abstractmethod
+    def _locked_run(self, raw_features):
+        """Device-specific evaluation of the model with the given raw features.
+
+        Must be called with self._lock held for read.
+        """
+        pass
+
+    @abc.abstractmethod
+    def _locked_shutdown(self, raw_features):
+        """Device-specific shutdown.
+
+        Must be called with self._lock held for write.
+        """
+        pass
+
+    @abc.abstractmethod
+    def _prepare_features(self, raw_features):
+        """Device-specific preparation of raw features.
+
+        Does not require a lock to be held.
+        """
+        pass
+
+
+class BasicSession(Session):
+    def __init__(self):
+        Session.__init__(self, tf.Session(graph=tf.Graph()))
+
+        with self._sess.graph.as_default():
+            self._feature_placeholder = tf.placeholder(
+                tf.float32, [None, go.N, go.N,
+                             features_lib.NEW_FEATURES_PLANES],
+                name='pos_tensor')
+
+    def _locked_shutdown(self):
+        pass
+
+    def _locked_load_model(self, path):
+        tf.reset_default_graph()
+
+        if path[-3:] == ".pb":
+            graph_def = tf.GraphDef()
+            with tf.gfile.FastGFile(path, 'rb') as f:
+                graph_def.ParseFromString(f.read())
+            with self._sess.graph.as_default():
+                self._outputs = tf.import_graph_def(
+                    graph_def,
+                    input_map={'pos_tensor': self._feature_placeholder},
+                    return_elements=['policy_output:0', 'value_output:0'])
+        else:
+            with self._sess.graph.as_default():
+                self._outputs = dual_net.model_inference_fn(
+                    self._feature_placeholder, training=False)
+                tf.train.Saver().restore(self._sess, path)
+
+    def _locked_run(self, features):
+        outputs = self._sess.run(self._outputs,
+                                 {self._feature_placeholder: features})
+        return outputs[0], outputs[1]
+
+    def _prepare_features(self, raw_features):
+        features = np.frombuffer(raw_features, dtype=np.int8)
+        features = features.reshape([-1, go.N, go.N,
+                                     features_lib.NEW_FEATURES_PLANES])
+        return features
+
+
+class TpuSession(Session):
+    def __init__(self, tpu_name, parallel_tpus, batch_size):
+        tpu = [tpu_name] if tpu_name else None
+        tpu_grpc_url = tf.contrib.cluster_resolver.TPUClusterResolver(
+            tpu=tpu).get_master()
+        sess = tf.Session(tpu_grpc_url)
+        Session.__init__(self, sess)
+
+        self._parallel_tpus = parallel_tpus
+        self._tpu_init = tf.contrib.tpu.initialize_system()
+        self._tpu_shutdown = tf.contrib.tpu.shutdown_system()
```

Comment:
oh weird, this must be an op.  You gotta admit, not knowing anything about it, it looks like you're just init'ing and then shutting down the system right here.

Reply:
good point, added a comment